packages/doenetml-worker-rust/lib-doenetml-core/src/core/general_state_var/string_state_var.test.rs

```diff
@@ -0,0 +1,545 @@
+    state_var.calculate_and_mark_fresh();
+
+    // we expect a value of "hello there"
+    // (a concatenation of the extending and the child)
+    // and came_from_default to be false
+    assert_eq!(state_var.get().clone(), String::from("hello there"));
+    assert_eq!(state_var.came_from_default(), false);
+
+    ////////////////////////////////////////////////////////////////
+    // Step 4: cannot invert to make the value "bye now"
+    ////////////////////////////////////////////////////////////////
+
+    // on the state variable view, record that we request the value be false
+    state_var_view.queue_update(String::from("bye now"));
+
+    let invert_result = state_var.invert(false);
+
+    assert!(invert_result.is_err());
+
+    ////////////////////////////////////////////////////////////////
+    // Step 5: change the extending and child to make it "bye now"
+    ////////////////////////////////////////////////////////////////
+
+    extend_var.set_value(String::from("bye "));
+    child_var.set_value(String::from("now"));
+
+    state_var.calculate_and_mark_fresh();
+
+    // now the value should be "bye now"
+    assert_eq!(state_var.get().clone(), "bye now");
+
+    ////////////////////////////////////////////////////////////////
+    // Step 6: verify changing essential does not affect the value
+    ////////////////////////////////////////////////////////////////
+
+    essential_var.set_value(String::from("ignore me"));
+
+    state_var.calculate_and_mark_fresh();
+
+    // the value should still be "bye now"
+    assert_eq!(state_var.get().clone(), "bye now");
+}
+
+/// Testing the case of string state variable based on an attribute
+/// For example, if we have a `<textInput prefill/>`
+/// its `prefill` state variable would follow this pattern.
+#[test]
+fn string_state_var_with_attribute() {
+    // create a boolean state variable requesting children
+    let mut state_var =
+        StringStateVar::new_from_attribute("my_attr", String::from("")).into_state_var();
+
+    //////////////////////////////////////////////////
+    // check that get the correct data queries
+    //////////////////////////////////////////////////
+    let queries = state_var.return_data_queries(None, 3);
+
+    // should return a query for "my_attr" with text children and an essential state variable
+    assert_eq!(
+        queries,
+        vec![
+            DataQuery::AttributeChild {
+                attribute_name: "my_attr",
+                match_profiles: vec![ComponentProfile::Text]
+            },
+            DataQuery::Essential
+        ]
+    );
+
+    // Note: from this point on, at the level of these tests where dependencies are specified manually,
+    // there is no difference in this state variable as the above ones with children,
+    // so we don't repeat the same steps
+}
+
+/// Testing the case of string state variable
+/// where we pass in a particular data query
```

Comment:
This isn't a good description. A good description will tell the future person who breaks the test why that test was there (Since future refactors may mean that some tests are not longer relevant and can be deleted).

Since I don't understand what this functionality is, I cannot suggest a good docstring, but maybe something like
```
When created, `StringStateVar`s specify where their values come from.
One possibility is that they come from another state var, which is sepcified
via a data query. 
...
```
Of course, some of this may be already written in the docstring of `StringStateVar`, so you could leave out much of the explanation and reference the corresponding doc. See https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html

Reply:
This is vague because I don't yet understand what we'd use `StringStateVar::new()` for either! We aren't using it anywhere, but I didn't feel ready to delete it in case we come with use for it soon.

Maybe rather than making this vague placeholder test, I'll just put a `TODO` in the source to either use the feature (and make a test that illustrates how it is used) or delete the feature. (If we want to make a string variable from a state variable, maybe we'll want a `StringStateVar::new_from_state_variable()` rather than this? It's not yet clear if allowing a generic data query is useful.)

Reply:
If we don't see a use case and it may be deleted, there probably shouldn't be a test for it.

Reply:
When you do those links to other docs, does it work for you in rust-analyzer? Mine just wants to send me to URLs.